datahub-web-react/src/components/entity/Entity.tsx

```diff
@@ -0,0 +1,58 @@
+import { EntityType } from '../../types.generated';
+
+export enum PreviewType {
+    /**
+     * A preview shown within the search experience
+     */
+    SEARCH,
+    /**
+     * A preview shown within the browse experience
+     */
+    BROWSE,
+    /**
+     * A generic preview shown within other entity pages, etc.
+     */
+    PREVIEW,
+}
+
+/**
+ * Base interface used for authoring DataHub Entities on the client side.
+ *
+ * <T> the generated GraphQL data type associated with the entity.
+ */
+export interface Entity<T> {
+    /**
+     * Corresponding GQL EntityType.
+     */
+    type: EntityType;
+
+    /**
+     * Returns whether the entity search is enabled
+     */
+    isSearchEnabled: () => boolean;
+
+    /**
+     * Returns whether the entity browse is enabled
+     */
+    isBrowseEnabled: () => boolean;
+
+    /**
+     * Returns the name of the entity as it appears in a URL, e.g. '/dataset/:urn'.
+     */
+    getPathName: () => string;
+
+    /**
+     * Returns the plural name of the entity used when displaying collections (search, browse results), e.g. 'Datasets'.
+     */
+    getCollectionName: () => string;
+
+    /**
+     * Renders the 'profile' of the entity on an entity details page.
+     */
+    renderProfile: (urn: string) => JSX.Element;
+
+    /**
+     * Renders a preview of the entity across different use cases like search, browse, etc.
+     */
+    renderPreview: (type: PreviewType, data: T) => JSX.Element;
```

Comment:
what do you think about renderPreview taking a default PreviewType of `PreviewType.Preview`?

Reply:
Yeah I like that!

Reply:
Update: Don't think it's possible to set default value on an interface parameter

> A parameter initializer is only allowed in a function or constructor implementation.ts(2371)

Reply:
@jjoyce0510 you could always just make type optional then in this API so that the implementors are forced to provide a default

Reply:
TypeScript types are just "removed" from the end Javascript. They have no impact on the code generation, only on type checking. So yeah, it makes sense you can't add default values to interfaces.

Also worth noting that even besides that point, TypeScript interfaces are duck-typed/structural. So it wouldn't make sense to have default values from that perspective either. e.g.

```
interface Example {
  foo(): number;
}

function giveExample(e: Example) {}

const obj = { foo() { return 2 } };
giveExample(obj); // obj did not explicitly implement Example, but it matches structurally, so this is ok.
```

Reply:
Also `optional` does not force a default; `undefined` is a implicit default.

```
interface Example {
  foo(param?: number): void;
}

class Impl implements Example {
  foo() {
    // ignoring an optional param is fine
  }
}

class Impl2 implements Example {
  foo(param = undefined) {
    // default of undefined is fine
    let n: number = param; // this is an error though, as param can be undefined
  }
}

class Impl3 implements Example {
  foo(param = 3) {
    let n: number = param; // not an error, TS knows this is always a number
  }
}

let e: Example = new Impl3();
e.foo(undefined);
```

Another way to think about this is the inheritance of functional types. `(): void` is a suptype of `(number?): void`.

Also worth noting that `?` sort of basically means `number|undefined`. So `undefined` is always a valid value for optionals, though at run time `undefined` is replaced with a default, if there is any.

Minor things that probably don't impact what you're doing, but worth understanding how the language works, imo :)